src/qibolab/instruments/qm.py

```diff
@@ -0,0 +1,598 @@
+import collections
+from types import SimpleNamespace
+
+import numpy as np
+from qibo.config import log, raise_error
+from qm.QuantumMachinesManager import QuantumMachinesManager
+
+from qibolab.instruments.abstract import AbstractInstrument
+from qibolab.result import ExecutionResult
+
+
+class QMOPX(AbstractInstrument):
+    """Instrument object for controlling Quantum Machines (QM) OPX controllers.
+
+    Playing pulses on QM controllers requires a ``config`` dictionary and a program
+    written in QUA language. The ``config`` file is generated in parts in the following places
+    in the ``register_*`` methods. The controllers, elements and pulses are all
+    registered after a pulse sequence is given, so that the config contains only
+    elements related to the participating qubits.
+    The QUA program for executing an arbitrary qibolab ``PulseSequence`` is written in
+    ``play`` and ``play_pulses`` and executed in ``execute_program``.
+
+    Args:
+        name (str): Name of the instrument instance.
+        address (str): IP address and port for connecting to the OPX instruments.
+
+    Attributes:
+        is_connected (bool): Boolean that shows whether instruments are connected.
+        manager (:class:`qm.QuantumMachinesManager.QuantumMachinesManager`): Manager object
+            used for controlling the QM OPXs.
+        config (dict): Configuration dictionary required for pulse execution on the OPXs.
+        time_of_flight (optional,int): Time of flight used for hardware signal integration.
+        smearing (optional,int): Smearing used for hardware signal integration.
+    """
+
+    def __init__(self, name, address):
+        # QuantumMachines manager is instantiated in ``platform.connect``
+        self.name = name
+        self.address = address
+        self.manager = None
+        self.is_connected = False
+
+        self.time_of_flight = 0
+        self.smearing = 0
+        # copied from qblox runcard, not used here yet
+        # hardware_avg: 1024
+        # sampling_rate: 1_000_000_000
+        # repetition_duration: 200_000
+        # minimum_delay_between_instructions: 4
+
+        # Default configuration for communicating with the ``QuantumMachinesManager``
+        self.config = {
+            "version": 1,
+            "controllers": {},
+            "elements": {},
+            "pulses": {},
+            "waveforms": {},
+            "digital_waveforms": {
+                "ON": {"samples": [(1, 0)]},
+            },
+            "integration_weights": {},
+            "mixers": {},
+        }
```

Comment:
Also this would be better as its own class, even specific to QM. You can also call it just `Configs` and place it in this module.

[In general](https://peps.python.org/pep-0020/):
> Namespaces are one honking great idea -- let's do more of those!

So, you don't have to use `qm.QMConfigs`. Since it is already part of the `qm` module, the `QM` in the class name is redundant.
In principle, you could do the same for this class, and call it just `OPX`, since it will `qm.OPX` (but here I understand if you want to keep the prefix)

Reply:
> Also this would be better as its own class, even specific to QM. You can also call it just `Configs` and place it in this module.

Just to explain, in case this `config` looks weird, it is required in order to execute programs on the QM hardware. Essentailly it is how we transfer data from the host machine to the hardware. Indeed the code could look better if we define this using a `Config` class but in the end we would need a method that dumps the `Config` object to a dictionary in order to execute programs using QM libraries.

Reply:
> but in the end we would need a method that dumps the `Config` object to a dictionary in order to execute programs using QM libraries.

This I understand, and it is generally a useful thing. Luckily, `dataclass`es provide them :D
https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

The main limitation is that they are not going to serialize and deserialize (i.e. transform to and from a `dict`) nested `dataclass`es.
For this we would need `attrs` with its companion `cattrs`, or `pydantic`, as said [below](). For a comparison of the various options I suggest reading
https://www.attrs.org/en/stable/why.html
But it should not be a concern for `Configs` at the moment, we should just keep it plain and use `dataclass` only for the first level, keeping dictionaries inside.